neo/Consensus/ConsensusContext.cs

```diff
@@ -123,8 +143,10 @@ public ConsensusPayload MakeChangeView()
         {
             return MakeSignedPayload(new ChangeView
             {
+                // View number is ignored for ChangeView, and for recovery we need it to be predicatable, so set as 0.
```

Comment:
Maybe it can be done before Regeneration, like we did in #426.
`InitializeConsensus(message.ViewNumber);`

Reply:
I'm not following you. This `ViewNumber` passed in change view is never used. Only the `NewViewNumber` is used.

Reply:
I don't think I follow what you are trying to say here. The `ViewNumber` passed with `ChangeView` is never used, only the `NewViewNumber` is ever used.

Reply:
#426 does not use the `ViewNumber` from a `ChangeView` message either for anything that I can tell, it also only uses the `NewViewNumber` in the message.

Reply:
We handle it here: https://github.com/neo-project/neo/blob/6cfcc8f3c4ca919e029099bb90de1a6df659ed5d/neo/Consensus/ConsensusService.cs#L407

Reply:
That is the Regeneration message's `ViewNumber`, different than the `ChangeView` `ViewNumber` that you are commenting on here.

Reply:
This code does that very same thing also here:

https://github.com/neo-project/neo/blob/17ba1eb95db7d953492a34617ee7bf8a596c10eb/neo/Consensus/ConsensusService.cs#L404

Reply:
The use is the same, for setting the view, I will take another look.

Reply:
@jsolman, you resolved the conversion two times. Please do not Resolve until under discussion.

Here in this PR, it looks like that Change View need to be verified.
On the other hand, in the previous PR the state commit being achieved by one state was enough to prove that.

Reply:
@vncoelho your comment doesn't make sense. The issue on the line in question you are commenting on is resolved. There isn't an issue with it. If you want to discuss something else in the PR you should open another message on the line you intend to dispute.

> Here in this PR, it looks like that Change View are only set after verified.

`ChangeView` are saved in this PR immediately when received; what do you mean by "only set after verified" also that discussion doesn't have anything to do with this line of code.

> If a node sends a Regeneration with a wrong View it is very easy to check if that was Malicious behavior

This code does check it when it receives recovery messages it ensures that signatures for change views for >= M nodes have are present in the recovery message.

I will not resolve this any more, I will wait for you to do it, but I don't see any real issue you are pointing out here that is unresolved.

Reply:
I suppose as an optimization if it at the time the recovery message is created there are >= M Preparations in the view it can avoid saving the `ChangeViewWitnessInvocationScripts` and the `ChangeViewTimestamps` in the recovery message, but it is just a small optimization to save space. I may make that optimization though.

Reply:
I made changes to the recovery to be slightly more optimal, so it now only regenerates the `ChangeView` messages if necessary (in the case there are not enough valid preparations present).

Reply:
@vncolho Is there something still unresolved here related to this conversation?

Reply:
I think that last commits https://github.com/neo-project/neo/pull/579/commits/95f28c0321a43c33d50af072d18d4ef3ba4bfb29 and https://github.com/neo-project/neo/pull/579/commits/4a3eb82e05151d6f96fddb5f3f4ede0af2717f31 added the necessary feature, right?
I will take another look and resolve it.

Reply:
Yes for this one, I think it can be resolved now. for the other two I have one more change coming that should resolve your last concerns.

Reply:
Oh yeah, I mentioned earlier I wouldn't be the one to resolve this one. @vncoelho can you please resolve this one now unless you have more concerns?

Reply:
Done. If necessary we open another ones considering the recent commits.